### PR TITLE
controllers: fix miscalculation of RunningDuration when killing job

### LINKED ISSUE #3715: Garbage Collector recalculates `job.Status.RunningDuration` after restarting controller
### Description

When restarting a `volcano-controllers-xxxx` pod, the Garbage Collector seems to check if previously completed jobs are ready for cleanup and "kills" them again, thus resulting in it updating the `job.Status.RunningDuration` variable, even though the job had already been completed before.

I use the job duration variable to indicate to a user how long a task has taken after it was completed, which ends up changing every time the controller is restarted due to some issue.

<details>
<summary>Controller logs</summary>

```
I0910 13:24:49.411220       1 job_controller.go:269] worker 2 start ......
I0910 13:24:49.411222       1 job_controller.go:269] worker 1 start ......
I0910 13:24:49.411236       1 job_controller.go:269] worker 0 start ......
I0910 13:24:49.411234       1 job_controller.go:320] Try to handle request <Queue: , Job: c774da7d-aff4-42cb-9eb0-f59912acecbb/test-log-spam, Task:, Event:OutOfSync, ExitCode:0, Action:, JobVersion: 0>
I0910 13:24:49.411234       1 job_controller.go:320] Try to handle request <Queue: , Job: c774da7d-aff4-42cb-9eb0-f59912acecbb/test-workload-4, Task:, Event:OutOfSync, ExitCode:0, Action:, JobVersion: 0>
I0910 13:24:49.411296       1 job_controller.go:337] Execute <SyncJob> on Job <c774da7d-aff4-42cb-9eb0-f59912acecbb/test-workload-4> in <Completed> by <*state.finishedState>.
I0910 13:24:49.411295       1 job_controller.go:337] Execute <SyncJob> on Job <c774da7d-aff4-42cb-9eb0-f59912acecbb/test-log-spam> in <Completed> by <*state.finishedState>.
I0910 13:24:49.411306       1 job_controller_actions.go:48] Killing Job <c774da7d-aff4-42cb-9eb0-f59912acecbb/test-workload-4>, current version 8
I0910 13:24:49.411307       1 job_controller_actions.go:48] Killing Job <c774da7d-aff4-42cb-9eb0-f59912acecbb/test-log-spam>, current version 8
I0910 13:24:49.411351       1 job_controller_actions.go:122] Running duration is 20h3m37.41134596s
I0910 13:24:49.411351       1 job_controller_actions.go:122] Running duration is 21h8m12.411345963s
W0910 13:24:49.421976       1 warnings.go:70] unknown field "spec.tasks[0].template.metadata.creationTimestamp"
I0910 13:24:49.423087       1 garbagecollector.go:124] Updating job c774da7d-aff4-42cb-9eb0-f59912acecbb/test-log-spam
I0910 13:24:49.423102       1 garbagecollector.go:132] Add job c774da7d-aff4-42cb-9eb0-f59912acecbb/test-log-spam to cleanup
I0910 13:24:49.423117       1 garbagecollector.go:191] Checking if Job c774da7d-aff4-42cb-9eb0-f59912acecbb/test-log-spam is ready for cleanup
W0910 13:24:49.423088       1 warnings.go:70] unknown field "spec.tasks[0].template.metadata.creationTimestamp"
I0910 13:24:49.423129       1 garbagecollector.go:290] Found Job c774da7d-aff4-42cb-9eb0-f59912acecbb/test-log-spam finished at 2024-09-09 17:23:44 +0000 UTC, remaining TTL 51h58m54.57687801s since 2024-09-10 13:24:49.42312199 +0000 UTC, TTL will expire at 2024-09-12 17:23:44 +0000 UTC
I0910 13:24:49.423814       1 garbagecollector.go:124] Updating job c774da7d-aff4-42cb-9eb0-f59912acecbb/test-workload-4
I0910 13:24:49.423828       1 garbagecollector.go:132] Add job c774da7d-aff4-42cb-9eb0-f59912acecbb/test-workload-4 to cleanup
I0910 13:24:49.423837       1 garbagecollector.go:191] Checking if Job c774da7d-aff4-42cb-9eb0-f59912acecbb/test-workload-4 is ready for cleanup
I0910 13:24:49.423846       1 garbagecollector.go:290] Found Job c774da7d-aff4-42cb-9eb0-f59912acecbb/test-workload-4 finished at 2024-09-09 16:19:24 +0000 UTC, remaining TTL 50h54m34.576158806s since 2024-09-10 13:24:49.423841194 +0000 UTC, TTL will expire at 2024-09-12 16:19:24 +0000 UTC
```

</details>

### Steps to reproduce the issue

1. Create a vcjob and wait for it to complete
2. Observe `Running Duration` in `kubectl describe vcjob ...`
3. Restart the Volcano controller
4. Observe `Running Duration` in `kubectl describe vcjob ...` again, the value has changed even though the job was already completed

### Describe the results you received and expected

If I understand it correctly, `Running Duration` should indicate how long a job took from start until completion, so this variable shouldn't change after the controller is restarted.

### What version of Volcano are you using?

1.8.0

### Any other relevant information

I tracked down the issue to [this line](https://github.com/volcano-sh/volcano/blob/master/pkg/controllers/job/job_controller_actions.go#L122) in the Garbage Collector which should probably use the job's end time instead of the current time.

However, I'm not entirely sure `killJob` should be called at all given that it has already been killed before.

### PR DESCRIPTION
**Pull request by matbme**
When restarting a controller pod, it calls `killJob` on completed jobs which will erroneously change the `RunningDuration` value for said job, even though it had already been completed. This PR fixes the issue by using the last transition time instead of current time to determine total running duration inside `killJob`.

Closes #3715.

### CODE REVIEW

**Review by hwdef — COMMENTED** [review]
/ok-to-test
/lgtm

File: pkg/controllers/job/job_controller_actions.go [hunk 1]
```diff
@@ -129,6 +124,11 @@ func (cc *jobcontroller) killJob(jobInfo *apis.JobInfo, podRetainPhase state.Pha
 		}
 	}
 
+	// Update running duration
+	runningDuration := metav1.Duration{Duration: job.Status.State.LastTransitionTime.Sub(jobInfo.Job.CreationTimestamp.Time)}
```
Comment (Monokaix) [inline pkg/controllers/job/job_controller_actions.go:128]: Can `job.Status.State.LastTransitionTime` indicate the job complete time?

Comment (matbme) [reply pkg/controllers/job/job_controller_actions.go:128]: We modify `LastTransitionTime` inside the if statement above when we receive an update status. Given that we only call `killJob` when transitioning to a terminal state or when restarting the controller, `LastTransitionTime` will be either:

1. The current time if there was a state change, or
2. The time when we last transitioned states, which will be whenever `killJob` was originally called for that job.

Unless I'm missing some other scenario, `LastTransitionTime` doesn't change after a job has been killed, thus it should hold the time when it transitioned to a completed state.

Comment (Monokaix) [reply pkg/controllers/job/job_controller_actions.go:128]: seems no problem: )

**Comment by Monokaix** [comment]
Hi, please rebase master.

**Comment by lowang-bh** [comment]
Could you please paste a test result  or add an ut?

**Comment by matbme** [comment]
Here's a manual test I did with and without this PR applied:

```sh
# ---------- Without PR ----------

$ k describe jobs.batch.volcano.sh test-workload-1 | grep -B 11 'Running Duration'
Status:
  Conditions:
    Last Transition Time:  2024-09-17T19:25:22Z
    Status:                Pending
    Last Transition Time:  2024-09-17T19:25:54Z
    Status:                Running
    Last Transition Time:  2024-09-17T19:26:12Z
    Status:                Completing
    Last Transition Time:  2024-09-17T19:26:12Z
    Status:                Completed
  Min Available:           1
  Running Duration:        6m8.810943003s
  
$ k -n volcano-system rollout restart deployment volcano-controllers
deployment.apps/volcano-controllers restarted

$ k describe jobs.batch.volcano.sh test-workload-1 | grep 'Running Duration'
Status:
  Conditions:
    Last Transition Time:  2024-09-17T19:25:22Z
    Status:                Pending
    Last Transition Time:  2024-09-17T19:25:54Z
    Status:                Running
    Last Transition Time:  2024-09-17T19:26:12Z
    Status:                Completing
    Last Transition Time:  2024-09-17T19:26:12Z
    Status:                Completed
  Min Available:           1
  Running Duration:        8m36.103534833s
  
  
# ---------- With PR ----------

$ k describe jobs.batch.volcano.sh test-workload-1 | grep -B 11 'Running Duration'
Status:
  Conditions:
    Last Transition Time:  2024-09-17T19:25:22Z
    Status:                Pending
    Last Transition Time:  2024-09-17T19:25:54Z
    Status:                Running
    Last Transition Time:  2024-09-17T19:26:12Z
    Status:                Completing
    Last Transition Time:  2024-09-17T19:26:12Z
    Status:                Completed
  Min Available:           1
  Running Duration:        51s
  
$ k -n volcano-system rollout restart deployment volcano-controllers
deployment.apps/volcano-controllers restarted

$ k describe jobs.batch.volcano.sh test-workload-1 | grep -B 11 'Running Duration'
Status:
  Conditions:
    Last Transition Time:  2024-09-17T19:25:22Z
    Status:                Pending
    Last Transition Time:  2024-09-17T19:25:54Z
    Status:                Running
    Last Transition Time:  2024-09-17T19:26:12Z
    Status:                Completing
    Last Transition Time:  2024-09-17T19:26:12Z
    Status:                Completed
  Min Available:           1
  Running Duration:        51s
```

Note that `Running Duration` wasn't correct even before restarting the controllers. My guess is that it was accounting for the time the job was pending, but I'm not entirely sure why my PR "fixes" this.

Regarding unit tests, I originally intended to add a test for this scenario, but had trouble replicating this behavior inside a ut. Please let me know if you have any indicators on how I could achieve this (maybe a e2e would be more fitting).

**Comment by Monokaix** [comment]
Please squash to one commit: )

**Comment by matbme** [comment]
Done!

**Comment by hwdef** [comment]
@william-wang @Thor-wl @shinytang6 
Please trigger the CI

**Comment by matbme** [comment]
Hi! Any updates on the CI?

**Pull request merged**